apps/cowswap-frontend/src/modules/swap/hooks/useSwapConfirmManager.ts

```diff
@@ -1,75 +1,110 @@
 export interface SwapConfirmManager {
   setSwapError(swapErrorMessage: string): void
-  openSwapConfirmModal(tradeToConfirm: TradeGp): void
+  openSwapConfirmModal(tradeToConfirm: TradeGp, needsPermitSignature?: boolean): void
   acceptRateUpdates(tradeToConfirm: TradeGp): void
   closeSwapConfirm(): void
   sendTransaction(tradeToConfirm: TradeGp): void
   transactionSent(txHash: string): void
+  requestPermitSignature(): void
+  permitSigned(): void
 }
 
 export function useSwapConfirmManager(): SwapConfirmManager {
-  const [swapConfirmState, setSwapConfirmState] = useAtom(swapConfirmAtom)
+  const setSwapConfirmState = useSetAtom(swapConfirmAtom)
   const [isExpertMode] = useExpertModeManager()
 
   return useMemo(
     () => ({
       setSwapError(swapErrorMessage: string) {
-        const state = { ...swapConfirmState, swapErrorMessage }
-        console.debug('[Swap confirm state] setSwapError: ', state)
-        setSwapConfirmState(state)
+        setSwapConfirmState((prev) => {
+          const state = { ...prev, swapErrorMessage, attemptingTxn: false, permitSignatureState: undefined }
+          console.debug('[Swap confirm state] setSwapError: ', state)
+          return state
+        })
       },
       openSwapConfirmModal(tradeToConfirm: TradeGp) {
-        const state = {
+        const state: SwapConfirmState = {
           tradeToConfirm,
           attemptingTxn: false,
           swapErrorMessage: undefined,
           showConfirm: true,
           txHash: undefined,
+          permitSignatureState: undefined,
         }
         console.debug('[Swap confirm state] openSwapConfirmModal: ', state)
         setSwapConfirmState(state)
       },
       acceptRateUpdates(tradeToConfirm: TradeGp) {
-        const state = { ...swapConfirmState, tradeToConfirm }
-        console.debug('[Swap confirm state] acceptRateUpdates: ', state)
-        setSwapConfirmState(state)
+        setSwapConfirmState((prev) => {
+          const state = { ...prev, tradeToConfirm }
+          console.debug('[Swap confirm state] acceptRateUpdates: ', state)
+          return state
+        })
+      },
+      requestPermitSignature() {
+        setSwapConfirmState((prev) => {
+          const state: SwapConfirmState = { ...prev, permitSignatureState: 'requested' }
+          console.debug('[Swap confirm state] requestPermitSignature: ', state)
+          return state
+        })
+      },
+      permitSigned() {
+        setSwapConfirmState((prev) => {
+          if (prev.permitSignatureState === 'requested') {
+            // Move to signed state only if previous state was good
+            // Set to undefined otherwise and make the action a no-op
+            const state: SwapConfirmState = { ...prev, permitSignatureState: 'signed' }
+            console.debug('[Swap confirm state] permitSigned: ', state)
+            return state
+          }
+          console.debug('[Swap confirm state] permitSigned was a no-op')
+          return prev
+        })
```

Comment:
Call this after the signing step.
No-op when permit is not expected.

Reply:
Instead of no-op, made it reset the property to be more consistent and less confusing